utils/kubernetes/kubeutil.py

```diff
@@ -56,27 +60,107 @@ def __init__(self, instance=None):
-        self.cadvisor_url = '%s://%s:%d' % (self.method, self.kubelet_host, self.cadvisor_port)
+        # apiserver
         self.kubernetes_api_url = 'https://%s/api/v1' % (os.environ.get('KUBERNETES_SERVICE_HOST') or self.DEFAULT_MASTER_NAME)
 
+        # kubelet
+        self.ssl_settings = self._init_ssl_settings(instance)
+        self.kubelet_api_url = self._locate_kubelet(instance)
+        if not self.kubelet_api_url:
+            log.exception("Kubernetes check exiting, cannot run without access to kubelet.")
+            return
+        self.kubelet_host = self.kubelet_api_url.split(':')[1].lstrip('/')
+        self.pods_list_url = urljoin(self.kubelet_api_url, KubeUtil.PODS_LIST_PATH)
+        self.kube_health_url = urljoin(self.kubelet_api_url, KubeUtil.KUBELET_HEALTH_PATH)
+
+        # cadvisor
+        self.cadvisor_port = instance.get('port', KubeUtil.DEFAULT_CADVISOR_PORT)
+        self.cadvisor_url = '%s://%s:%d' % (self.method, self.kubelet_host, self.cadvisor_port)
         self.metrics_url = urljoin(self.cadvisor_url, KubeUtil.METRICS_PATH)
         self.machine_info_url = urljoin(self.cadvisor_url, KubeUtil.MACHINE_INFO_PATH)
-        self.pods_list_url = urljoin(self.kubelet_api_url, KubeUtil.PODS_LIST_PATH)
-        self.kube_health_url = urljoin(self.kubelet_api_url, 'healthz')
 
         # keep track of the latest k8s event we collected and posted
         # default value is 0 but TTL for k8s events is one hour anyways
         self.last_event_collection_ts = 0
 
+    def _init_ssl_settings(self, instance):
+        """
+        Extract SSL settings from the config.
+        """
+        ssl_settings = {}
+
+        client_crt = instance.get('kubelet_client_crt')
+        client_key = instance.get('kubelet_client_key')
+        if client_crt and client_key:
+            ssl_settings['kubelet_client_cert'] = (client_crt, client_key)
+
+        cert = instance.get('kubelet_cert')
+        if cert:
+            ssl_settings['verify'] = cert
+        else:
+            ssl_settings['verify'] = instance.get('kubelet_ssl_verify', DEFAULT_SSL_VERIFY)
+
+        return ssl_settings
+
+    def _locate_kubelet(self, instance):
+        """
+        Kubelet may or may not accept un-authenticated http requests.
+        If it doesn't we need to use its HTTPS API that may or may not
+        require auth.
+        """
+        host = os.environ.get('KUBERNETES_KUBELET_HOST') or instance.get("host")
+        if not host:
+            # if no hostname was provided, use the docker hostname if ssl
+            # validation is not required, the kubernetes hostname otherwise.
+            docker_hostname = self.docker_util.get_hostname()
+            if self.ssl_settings['verify']:
+                try:
+                    k8s_hostname = self.get_node_hostname(docker_hostname)
+                    host = k8s_hostname or docker_hostname
+                except Exception as ex:
+                    log.error(str(ex))
```

Comment:
`get_node_hostname` can raise if reaching the apiserver (connecting, validating its cert...) fails.

Right now it can also happen if the host filtering doesn't work and we can't pick which hostname to use for the local kubelet, but I'll actually make that a `log.error` since we can try and fallback to the docker hostname and fail later if cert validation doesn't like that.